rfd/0080-piv-private-key.md

```diff
@@ -0,0 +1,316 @@
+ - The hardware key is connected during the operation
+ - The hardware private key can still be found
+ - The hardware private keys' Touch challenges are passed (if applicable)
+
+#### PIV slot overwrite/reuse logic
+
+For PIV hardware keys, we will use slot `9a`, which is reserved for authentication. Before generating a new private key, Teleport clients will check if PIV slot `9a` is already in use by a Teleport client. If it is, then the client will reuse the existing private key instead of generating a new one. Otherwise, it will generate a new private key.
+
+#### Private key interface
+
+Currently, Teleport clients store a PEM encoded private key (`~/.tsh/keys/proxy/user`) for a login session. This PEM encoded private key is then unmarshalled, transformed, and parsed as needed during a client request.
+
+With a hardware private key, we only have access to a raw `crypto.PrivateKey`, and do not have sufficient information about the key to transform it into a `*rsa.PrivateKey` and marshal it into PKCS1 format. Instead, we need to alter Teleport clients to use `cyrpto.PrivateKey` by default. This will require altering the key interface (`lib/client/interfaces.go`) and its usage across `lib/client` and other relevant locations. `lib/utils/native` will also be updated to return `*rsa.PrivateKey` instead of its PEM encoded private and public keys.
+
+We also need a way for future Teleport Client requests to retrieve the correct `crypto.PrivateKey`. For RSA keys, we can continue to store them as PEM encoded keys in (`~/.tsh/keys/proxy/user`). For hardware private keys, we will instead store a fake PEM encoded private key which we can use to identity what device and slot to load the private key from.
+
+```
+-----BEGIN YUBIKEY PRIVATE KEY-----
+# PEM encoded `yubikey_serial_number+slot`
+-----END YUBIKEY PRIVATE KEY-----
+```
+
+#### Supported clients
+
+`tsh` and Teleport Connect will both support hardware private key login, and `tctl` will be able to use resulting login sessions. 
+
+Supporting hardware private key login through the WebUI needs to be investigated more thoroughly, but it poses a challenge since it is browser-based and does not have direct access to the user's local hardware keys.
+
+### UX
+
+The most notable UX change is that a user's login session will not be usable unless their hardware key is connected.
+
+If `private_key_policy: hardware_key_touch` is used, then every Teleport Client request will require touch (cached for 15 seconds). This will be handled by a touch prompt similar to the one used for MFA.
+
+### Additional considerations
+
+#### Database support
+
+`tsh db connect` uses raw RSA private key data to form connections. Since this cannot be supported with hardware private keys, users will instead need to use `tsh proxy db` to connect using a local proxy. Teleport Connect already uses `tsh proxy db` and will not be affected, but the WebUI may have an additional challeng to support datbase connections.
```

Comment:
```suggestion
`tsh db connect` uses raw RSA private key data to form connections. Since this cannot be supported with hardware private keys, users will instead need to use `tsh proxy db` to connect using a local proxy. Teleport Connect already uses `tsh proxy db` and will not be affected, but the WebUI may have an additional challeng to support database connections.
```